js/node/src/directml_load_helper.cc

```diff
@@ -5,37 +5,40 @@
 #include "common.h"
 #include "windows.h"
 
-extern "C" __declspec(dllexport) void ExportFunction()
+void LoadDirectMLDll(Napi::Env env)
 {
-}
+  DWORD pathLen = MAX_PATH;
+  wchar_t* path = new wchar_t[pathLen];
+  HMODULE moduleHandle = nullptr;
 
-std::string GetModuleDirectory()
-{
-  char path[MAX_PATH];
-  HMODULE hm = NULL;
+  GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS |
+                                 GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
+                             reinterpret_cast<LPCSTR>(&LoadDirectMLDll), &moduleHandle);
 
-  if (!GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS |
-                             GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
-                         (LPCSTR) &ExportFunction, &hm))
-  {
-    int ret = GetLastError();
-    fprintf(stderr, "GetModuleHandleEx failed, error = %d\n", ret);
-  }
+  DWORD getModuleFileNameResult = GetModuleFileNameW(moduleHandle, path, pathLen);
+  while (getModuleFileNameResult == 0 || getModuleFileNameResult == pathLen) {
```

Comment:
(minor) I'd move the `pathLen = pathLen * 2;` inside the `if` so it's clearer associated with the resize, because in the `getModuleFileNameResult == 0` case, there's no need to double the buffer given that's a hard failure anyway.

```
   if (ret == ERROR_INSUFFICIENT_BUFFER && pathLength < 32768) {
      pathLength = pathLength * 2;
      path.resize(pathLength);
      getModuleFileNameResult = GetModuleFileNameW(moduleHandle, path, pathLen);
    } else {
```

Reply:
No need for the `const_cast` - I just use `path.data()`.

Reply:
Nope.
![image](https://github.com/microsoft/onnxruntime/assets/7466130/6447549c-84e9-4a06-b36c-19490aec15fa)
![image](https://github.com/microsoft/onnxruntime/assets/7466130/f4389ac3-bfc4-401d-a576-c42b861deb26)

Reply:
🤨 Huh, how weird There's definitely a non-const [`data()` overload](https://en.cppreference.com/w/cpp/string/basic_string/data) in C++17, and it builds fine for me locally in ONNX Runtime with Visual Studio 2022 (albeit a different file).

![image](https://github.com/microsoft/onnxruntime/assets/1809166/03acc803-b379-415d-999c-a3a171ccb122)

🤷‍♂️ Ok.

Reply:
I'm not building it with VS directly but with a build script in the root directory ` .\build.bat --config Release --build_shared_lib --parallel --build_nodejs --use_dml --skip_tests --use_cuda --cuda_home "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8" --cudnn_home "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8" --use_tensorrt --tensorrt_home "C:\Program Files\NVIDIA GPU Computing Toolkit\TensorRT-8.5.1.7"`

Reply:
🤷‍♂️ Well, keeping that one, we don't *this* one, right?

`auto libraryLoadResult = LoadLibraryW(const_cast<wchar_t*>(path.c_str()));`

Since it's only an input parameter and not output:

```
HMODULE LoadLibraryW(
  [in] LPCWSTR lpLibFileName
);
```